File: LeetCode/practice-cpp/LeetCode-42.cpp
// 接雨水
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> leftHeight = height;
        for (int i = 1; i < height.size(); ++i) {
            leftHeight[i] = max(leftHeight[i - 1], height[i]);
        }

        vector<int> rightHeight = height;
        for (int i = height.size() - 2; i >= 0; --i) {
            rightHeight[i] = max(rightHeight[i + 1], height[i]);
        }

        int trapSum = 0;
        for (int i = 0; i < height.size(); ++i) {
            trapSum += min(leftHeight[i], rightHeight[i]) - height[i];
        }
        return trapSum;
    }
};

File: LeetCode/practice-rust/LeetCode-42.rs
// 接雨水
use std::cmp;

struct Solution;

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        if height.is_empty() { return 0; }

        let mut max_right = 0;
        let from_right: Vec<i32> = height.iter().rev()
            .map(|&num|{
                max_right = cmp::max(max_right, num);
                max_right
            }).collect();        

        let mut max_left = 0;
        height.iter().zip(from_right.iter().rev())
            .map(|(&num, &max_right)|{
                max_left = cmp::max(max_left, num);
                cmp::min(max_left, max_right) - num
            }).sum()
    }
}

fn main () {
    let l1 = Solution::trap(vec![0,1,0,2,1,0,1,3,2,1,2,1]);
    println!("{:?}", l1);
}